sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/percentiles.scala

```diff
@@ -325,3 +339,133 @@ case class Percentile(
     frequencyExpression = newThird
   )
 }
+
+/**
+ * Return a percentile value based on a continuous distribution of
+ * numeric or ansi interval column at the given percentage (specified in ORDER BY clause).
+ * The value of percentage must be between 0.0 and 1.0.
+ */
+case class PercentileCont(left: Expression, right: Expression)
+  extends AggregateFunction
+  with RuntimeReplaceableAggregate
+  with ImplicitCastInputTypes
+  with BinaryLike[Expression] {
+  private lazy val percentile = new Percentile(left, right)
+  override def replacement: Expression = percentile
+  override def nodeName: String = "percentile_cont"
+  override def inputTypes: Seq[AbstractDataType] = percentile.inputTypes
+  override protected def withNewChildrenInternal(
+      newLeft: Expression, newRight: Expression): PercentileCont =
+    this.copy(left = newLeft, right = newRight)
+}
+
+/**
+ * The Percentile aggregate function returns the percentile(s) based on a discrete distribution of
+ * numeric column `expr` at the given percentage(s) with value range in [0.0, 1.0].
+ *
+ * Because the number of elements and their partial order cannot be determined in advance.
+ * Therefore we have to store all the elements in memory, and so notice that too many elements can
+ * cause GC paused and eventually OutOfMemory Errors.
+ */
+case class PercentileDisc private(
+    child: Expression,
+    percentageExpression: Expression,
+    frequencyExpression: Expression,
+    reverse: Boolean = false,
+    mutableAggBufferOffset: Int = 0,
+    inputAggBufferOffset: Int = 0) extends PercentileBase {
+
+  def this(child: Expression, percentageExpression: Expression) = {
+    this(child, percentageExpression, Literal(1L))
+  }
+
+  def this(child: Expression, percentageExpression: Expression, reverse: Boolean) = {
+    this(child, percentageExpression, Literal(1L), reverse)
+  }
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    val defaultCheck = super.checkInputDataTypes()
+    if (defaultCheck.isFailure) {
+      defaultCheck
+    } else {
+      assert(frequencyExpression == Literal(1L))
+      TypeCheckSuccess
+    }
+  }
+
+  override def prettyName: String = "percentile_disc"
+
+  /**
+   * This function has been based upon similar function from H2
+   * `org.h2.expression.aggregate.Percentile.getValue`.
+   */
+  override protected def getPercentiles(sortedCounts: Seq[(AnyRef, Long)]): Seq[Double] = {
+    val passedSortedCounts = if (reverse) {
+      sortedCounts.reverse
+    } else {
+      sortedCounts
+    }
+    val maxPosition = passedSortedCounts.last._2 - 1
+
+    percentages.map { percentile =>
+      getPercentile(passedSortedCounts, maxPosition, percentile)
+    }
+  }
+
+  private def getPercentile(
+      sortedCounts: Seq[(AnyRef, Long)], maxPosition: Long, percentile: Double): Double = {
+    val position = maxPosition * percentile
+    var lower = position.floor.toLong
+    val factor = position - lower
```

Comment:
shouldn't `factor` always be positive?

Reply:
For example: inputs:[10, 20, 30], percentile:0.5. Then position=1.0 and lower=1